README.md
# SCSAR Package

SCSAR is an R package specifically designed to facilitate the process of 
cell-type annotation for single-cell RNA-seq data analysis. It takes differential gene expression information for each cell cluster as input. Leveraging the 
comprehensive cell type marker gene data available in the 
[CellMarker 2.0](http://bio-bigdata.hrbmu.edu.cn/CellMarker/) and 
[Cell Taxonomy](https://ngdc.cncb.ac.cn/celltaxonomy/) databases, SCSAR employs 
a scoring algorithm to accurately and reliably assign cell-type annotations.

The Python version of SCSAR is [SCSA](https://github.com/bioinfo-ibms-pumc/SCSA). Compared to SCSA, SCSAR further includes Cell Taxonomy
as the marker gene resource, and enhances the scoring scheme by introducing a 
penalty system that penalizes cell types containing marker genes that are not 
overexpressed in the given cluster. This refinement leads to improved accuracy 
in cell-type identification. In addition, SCSAR provides the user with the top 
scoring markers for each cluster, enabling them to verify the results. Finally, 
SCSAR is more user-friendly for researchers performing single-cell sequencing 
analysis in R. Users no longer need to switch to Python for cell-type 
annotation, simplifying the workflow.


## Installation

To install the SCSAR package, you can use the `remotes` package:

``` r
if (!requireNamespace("remotes", quietly=TRUE))
    install.packages("remotes")
remotes::install_github("LanyingWei/SCSAR")
```

## Vignette

For more details, please refer to the 
[online vignette](https://htmlpreview.github.io/?https://github.com/LanyingWei/SCSAR/blob/main/vignettes/SCSAR.html).


NEWS
Changes in version 0.99.0 (2023-06-15)
+ Submitted to Github


R/data.R
#' PBMC 3k Differentially Expressed Genes (DEGs)
#'
#' The `pbmc3kDEGs` data is a data frame containing the differentially expressed genes (DEGs) for each cluster in the peripheral blood mononuclear cell (PBMC) dataset. The dataset consists of 3k PBMCs from a Healthy Donor and is freely available from 10x Genomics. The data can be accessed at: \url{https://www.10xgenomics.com/resources/datasets/3-k-pbm-cs-from-a-healthy-donor-1-standard-1-0-0}.
#'
#' The DEGs were identified through normalization, clustering, and differential gene expression analyses using the R package Seurat 4.1. Clustering of cells was performed using the implemented community identification method in the 'FindClusters' function. The specific marker genes of cell clusters (the DEGs for a specific cluster) were identified by comparing cells from that cluster to all other cells using the Seurat 'FindMarkers' function.
#'
#' @format A data frame with the following columns:
#'   \describe{
#'     \item{cluster}{Cluster identifier.}
#'     \item{gene}{Gene symbol.}
#'     \item{p_val}{p-value for differential expression.}
#'     \item{avg_log2FC}{Average log2 fold change.}
#'   }
#'
#' @usage data(pbmc3kDEGs)
#'
"pbmc3kDEGs"

R/showTables.R
#' Show CellMarker Database Tissue Information
#'
#' This function displays a searchable table with the tissue information and the corresponding species from the CellMarker 2.0 database.
#'
#' @return an HTML widget to display tabular data.
#' @export
#'
#' @examples
#' showCellMarkerTissue()
#'
#' @importFrom DT datatable
#'
showCellMarkerTissue <- function() {
    DB <- cellMarkerDB
    DB <- DB[, c("species", "tissue_class", "tissue_type")]
    DB <- unique(DB)
    rownames(DB) <- seq(length = nrow(DB))
    resTable <- DT::datatable(DB)
    return(resTable)
}

#' Show Cell Taxonomy Database Species Information
#'
#' This function displays a searchable table with the species information and the number of corresponding cell types from the Cell Taxonomy database.
#' 
#' @return an HTML widget to display tabular data.
#' @export
#'
#' @examples
#' showCellTaxoSpecies()
#'
#' @importFrom DT datatable
#' @importFrom dplyr group_by summarise
#'
showCellTaxoSpecies <- function() {
    DB <- cellTaxoDB
    DB <- DB[, c("species", "cell_name")]
    DB <- DB %>% 
        group_by(species) %>% 
        summarise(count = n_distinct(cell_name))
    colnames(DB)[2] <- "number_of_cell_types"
    DB <- DB[order(DB$number_of_cell_types, decreasing = TRUE), ]
    DB <- as.data.frame(DB)
    resTable <- DT::datatable(DB)
    return(resTable)
}


R/cellMarkerSupport.R
#' Generate Marker Gene Evidence Matrix (CellMarker 2.0 DB)
#'
#' This internal function generates a marker gene evidence matrix based on the CellMarker 2.0 database and given filtering criteria.
#'
#' @param markerDB A data frame containing the marker gene database.
#' @param species A character vector specifying the species to consider. Currently it supports "human" and "mouse". Use "all" to include all species.
#' @param tissueClass A character vector specifying the broad tissue classes to consider (case insensitive). Default is "all" that includes all tissue classes. To explore the available tissue classes, use the \code{showCellMarkerTissue()} function.
#' @param tissueType A character vector specifying the specific tissue types to consider within the chosen tissue class (case insensitive). Default is "all" that includes all tissue types. "tissueType" represents the specific subtypes within "tissueClass", For instance, when considering blood as "tissueClass", examples of blood tissue types could be "Peripheral blood", "Serum", "Blood vessel", and so on. To explore the available tissue types, use the \code{showCellMarkerTissue()} function.
#' @param cellState A character vector specifying the cell states to consider. Valid options are "normal" and "cancer". Default is both.
#' @param markerSource A character vector specifying the marker gene sources to consider. Valid options are "experiment", "review", "single-cell sequencing", and "company". Default is all.
#' @param logBase The base value for logarithmic transformation. Default is the Euler's number.
#' @param pseudoCount The pseudo count value added to the raw counts before logarithmic transformation. Default is 1.
#'
#' @return A matrix containing marker gene evidence levels. Row names represent cell types, and column names represent genes. The higher the value, the stronger the evidence that the gene is a marker for the cell type.
#' 
#' @importFrom dplyr "%>%" n_distinct

cellMarkerSupport <- function(markerDB,
                              tissueClass="all",
                              tissueType="all",
                              species="human",
                              cellState=c("normal", "cancer"),
                              markerSource=c("Experiment", 
                                             "Review",
                                             "Single-cell sequencing",
                                             "Company"),
                              logBase=exp(1),
                              pseudoCount=1) {
    
    stopifnot(!missing(markerDB))
    stopifnot(!missing(species))
    
    speciesAll <- tolower(sort(unique(markerDB$species)))
    tissueClassAll <- tolower(sort(unique(markerDB$tissue_class)))
    tissueTypeAll <- tolower(sort(unique(markerDB$tissue_type)))
    
    markerDB <- markerDB[!is.na(markerDB$species) &
                             !is.na(markerDB$tissue_class) &
                             !is.na(markerDB$tissue_type) &
                             !is.na(markerDB$cell_type) &
                             !is.na(markerDB$Symbol) &
                             !is.na(markerDB$PMID), ]
    
    markerDB$species <- tolower(markerDB$species)
    markerDB$tissue_class <- tolower(markerDB$tissue_class)
    markerDB$tissue_type <- tolower(markerDB$tissue_type)
    markerDB$marker_source <- tolower(markerDB$marker_source)
    species <- tolower(species)
    tissueClass <- tolower(tissueClass)
    tissueType <- tolower(tissueType)
    markerSource <- tolower(markerSource)
    
    stopifnot(markerSource %in% c("experiment", "review",
                                  "single-cell sequencing","company"))
    
    stopifnot(cellState %in% c("normal", "cancer"))
    
    if("all" %in% tolower(species)) {
        species <- speciesAll
    }
    
    if("all" %in% tolower(tissueClass)) {
        tissueClass <- tissueClassAll
    }
    
    if("all" %in% tolower(tissueType)) {
        tissueType <- tissueTypeAll
    }
    
    if (!all(species %in% speciesAll)) {
        stop("Species", 
             paste(species[!species %in% speciesAll], collapse = ", "), 
             " not in the marker database.",
             " To explore the available species,",
             " use the showCellMarkerTissue() function.")
    }
    
    if (!all(tissueClass %in% tissueClassAll)) {
        stop("Tissue Class", paste(tissueClass[!tissueClass %in% tissueClassAll], collapse = ", "), 
             " not in the marker database.",
             " To explore the available tissue classes,",
             " use the showCellMarkerTissue() function.")
    }
    
    if (!all(tissueType %in% tissueTypeAll)) {
        stop("Tissue Type", paste(tissueType[!tissueType %in% tissueTypeAll], collapse = ", "), 
             " not in the marker database.",
             " To explore the available tissue types,",
             " use the showCellMarkerTissue() function.")
    }
    
    cellState[cellState %in% "normal"] <- "Normal cell"
    cellState[cellState %in% "cancer"] <- "Cancer cell"
    
    markerDB2Use <- markerDB[markerDB$marker_source %in% markerSource &
                                 markerDB$species %in% species &
                                 markerDB$tissue_class %in% tissueClass &
                                 markerDB$tissue_type %in% tissueType &
                                 markerDB$cell_type %in% cellState, 
                             c("species", "tissue_class", "tissue_type", "cell_type",
                               "cell_name", "Symbol", "PMID")]
    
    markerDB2Use$cellInfo <- paste(markerDB2Use$species,
                                   markerDB2Use$tissue_class,
                                   markerDB2Use$tissue_type,
                                   markerDB2Use$cell_type,
                                   markerDB2Use$cell_name, sep = ":")
    
    pmidCounts <- markerDB2Use %>% 
        group_by(cellInfo, Symbol) %>% 
        summarise(count = n_distinct(PMID))
    
    # Spread the data into a matrix format
    pmidCounts <- pmidCounts  %>% 
        tidyr::spread(Symbol, count)
    cellInfo <- pmidCounts$cellInfo
    
    pmidCounts <- as.matrix(pmidCounts[, -1])
    row.names(pmidCounts) <- cellInfo
    pmidCounts[is.na(pmidCounts)] <- 0
    
    markerEvi <- log(pmidCounts + pseudoCount, base = logBase)
    
    pmidCountsCellInfo <- markerDB2Use %>% 
        group_by(cellInfo) %>% 
        summarise(count = n_distinct(PMID))
    
    pmidCountsCellInfo <- 
        pmidCountsCellInfo[match(rownames(pmidCounts), 
                                 pmidCountsCellInfo$cellInfo), ]
    
    markerCountsCellInfo <- markerDB2Use %>% 
        group_by(cellInfo) %>% 
        summarise(count = n_distinct(Symbol))
    
    markerCountsCellInfo <- 
        markerCountsCellInfo[match(rownames(pmidCounts), 
                                   markerCountsCellInfo$cellInfo), ]
    
    cellInfoEvi <- log(pmidCountsCellInfo$count * markerCountsCellInfo$count +
                           pseudoCount, base = logBase)
    names(cellInfoEvi) <- rownames(pmidCounts)
    
    weight <- markerEvi * cellInfoEvi
    
    return(weight)
}

R/cellTypeIdentifier.R
#' Cell Type Annotation
#'
#' This is an internal function that assigns cell types to cell clusters of single-cell RNA-seq data based on differential gene expression information and marker gene evidence.
#'
#' @param clusterDEGs A data frame containing information about differentially expressed genes for each cell cluster. It should have the following columns:
#' \describe{
#'   \item{cluster}{The cluster identifier or label.}
#'   \item{gene}{The gene symbol.}
#'   \item{p_val}{The p-value representing the significance of differential expression for the gene in the cluster.}
#'   \item{avg_log2FC}{The average log fold change indicating the magnitude of differential expression for the gene in the cluster.}
#' }
#' @param markerEviMat A matrix containing the evidence levels of marker genes. Rows represent cell types, and columns represent genes. The values in the matrix indicate the strength of evidence that a gene is a marker for a particular cell type. A higher value indicates stronger evidence.
#' @param clusters A character vector specifying the clusters to analyze. By default, all clusters are considered (set to "all").
#' @param cutoffLFC The log fold change cutoff value for the column "avg_log2FC" in the \code{clusterDEGs} data frame. Only cell type marker genes with an average log fold change greater than or equal to this value are considered as supporting evidence for the corresponding cell type. Default is 1, which corresponds to a two-fold change in average.
#' @param cutoffPval The p-value cutoff for the column "p_val" in the \code{clusterDEGs} data frame. Only cell type marker genes with a p-value less than or equal to this value are considered as supporting evidence for the corresponding cell type. Default is 0.05.
#' @param topCell The maximum number of top-ranked cell types to report. Default is 5.
#' @param topMarker The maximum number of top-ranked marker genes contributing to the cell type annotation to report for each cell type. Default is 5.
#' @param panelty The penalty weight assigned to cell type marker genes that are not a detected marker of the cluster. A "detected marker" means that the gene meets both the cutoffPval and paneltyCutoffLFC thresholds. A higher absolute value of the penalty means a more severe punishment for cell types with some of the marker genes that are not overexpressed in the cluster. The default value is -0.1.
#' @param paneltyCutoffLFC The log fold change cutoff value for including genes in the penalty calculation. Cell types with marker genes having an average log fold change less than this value are penalized. The default value is the same as "cutoffLFC". The value of "paneltyCutoffLFC" should be less than or equal to "cutoffLFC".
#'
#' @return A list containing the annotation results for each cluster. Each element in the list corresponds to a cluster and includes a data frame. The data frame has annotated cell types as row names and two columns: "score" (representing the final score for the cell type) and "topMarkers" (containing the top marker genes and their scores).
#'
#' @importFrom stats sd

cellTypeIdentifier <- function(clusterDEGs, 
                               markerEviMat, 
                               clusters="all",
                               cutoffLFC=1,
                               cutoffPval=0.05,
                               topCell=5,
                               topMarker=5,
                               panelty=-0.1,
                               paneltyCutoffLFC=cutoffLFC) {
    
    stopifnot(!missing(clusterDEGs))
    stopifnot(!missing(markerEviMat))
    
    if("all" %in% tolower(clusters)) {
        clusters <- sort(unique(clusterDEGs$cluster))
    }
    
    res <- list()
    for (i in seq_along(clusters)) {
        cluster <- clusters[i]
        thisClusterDEGsAll <- clusterDEGs[clusterDEGs$cluster %in% cluster &
                                              clusterDEGs$gene %in% 
                                              colnames(markerEviMat), 
                                          c("gene", "p_val", "avg_log2FC")]
        thisClusterDEGs <- 
            thisClusterDEGsAll[thisClusterDEGsAll$p_val <= cutoffPval &
                                   thisClusterDEGsAll$avg_log2FC >= cutoffLFC,]
        
        if (nrow(thisClusterDEGs) == 0) {
            stop("For cluster ", clusters[i], ", ",
                 sum(thisClusterDEGsAll$p_val <= cutoffPval), 
                 " gene(s) meet the CutoffPval threshold, and ",   
                 sum(thisClusterDEGsAll$avg_log2FC >= cutoffLFC), 
                 " gene(s) meet the PaneltyCutoffLFC threshold.",
                 " There are no genes left after filtering for both criteria,",
                 " please adjust the thresholds and run again.")
        }
        
        
        clusterMarkers <- thisClusterDEGs$gene
        message("Estimating identity of cell cluster ", cluster,
                " based on ", length(clusterMarkers), 
                " cluster marker genes...")
        
        # Retain only cell types that have at least one marker gene that is DEG
        thisMarkerEviMat <- markerEviMat[, clusterMarkers, drop = FALSE]
        thisMarkerEviMat <- thisMarkerEviMat[rowSums(thisMarkerEviMat) > 0, , 
                                             drop = FALSE]
        
        # Remove genes that are not the marker gene for any remaining cell types
        thisMarkerEviMat <-  markerEviMat[rownames(thisMarkerEviMat), ]
        thisMarkerEviMat <- thisMarkerEviMat[, colSums(thisMarkerEviMat) > 0]
        
        # Initialize weight vector with penalty score
        weightVec <- rep(panelty, ncol(thisMarkerEviMat))
        names(weightVec) <- colnames(thisMarkerEviMat)
        
        # Assign weights for supporting marker genes
        weightVec[thisClusterDEGs$gene] <- thisClusterDEGs$avg_log2FC
        
        # For genes with avg_log2FC above paneltyCutoffLFC, set penalty to 0,
        # so that these genes neither add to nor lower the cell type score
        noPaneltyGene <- 
            thisClusterDEGsAll$gene[thisClusterDEGsAll$gene %in% 
                                        names(weightVec) &
                                        thisClusterDEGsAll$p_val <= cutoffPval &
                                        thisClusterDEGsAll$avg_log2FC >= 
                                        paneltyCutoffLFC &
                                        thisClusterDEGsAll$avg_log2FC < 
                                        cutoffLFC]
        weightVec[noPaneltyGene] <- 0
        
        score <- thisMarkerEviMat %*% weightVec
        colnames(score) <- "score"
        score <- score[order(score, decreasing = TRUE), , drop = FALSE]
        zscore <- (score - mean(score)) / sd(score)
        zscore <- zscore[zscore > 0, , drop = FALSE]
        zscore <- zscore[seq_len(min(topCell, nrow(zscore))), , drop = FALSE]
        zscore <- round(zscore, 2)
        
        markerGeneScore <- 
            t(t(thisMarkerEviMat[rownames(zscore), ]) * weightVec)
        markerGenes <- 
            apply(markerGeneScore, 1, function(x) {
                markerScore <- sort(x[x>0], decreasing = TRUE)
                markerScore <- 
                    markerScore[seq_len(min(topMarker, length(markerScore)))]
                paste(paste(names(markerScore), 
                            round(markerScore, 0), sep = " "), collapse = ", ")
            })
        resInfo <- as.data.frame(zscore)
        resInfo$topMarkers <- markerGenes
        res[[paste("cluster", cluster)]] <- resInfo
    }
    message("Done.")
    return(res)
}


R/cellTaxonomySupport.R
#' Generate Marker Gene Evidence Matrix (Cell Taxonomy DB)
#'
#' This internal function generates a marker gene evidence matrix based on the Cell Taxonomy database and given filtering criteria.
#'
#' @param markerDB A data frame containing the marker gene database.
#' @param species A character vector specifying the species to consider. Currently it contains data on 34 species. Use "all" to include all species. Default is "Homo sapiens". To explore the available species, use the \code{showCellTaxoSpecies()} function.
#' @param logBase The base value for logarithmic transformation. Default is the Euler's number.
#' @param pseudoCount The pseudo count value added to the raw counts before logarithmic transformation. Default is 1.
#'
#' @return A matrix containing marker gene evidence levels. Row names represent cell types, and column names represent genes. The higher the value, the stronger the evidence that the gene is a marker for the cell type.
#' 
#' @importFrom dplyr "%>%" n_distinct

cellTaxonomySupport <- function(markerDB,
                                species="Homo sapiens",
                                logBase=exp(1),
                                pseudoCount=1) {
    
    stopifnot(!missing(markerDB))
    
    speciesAll <- tolower(sort(unique(markerDB$species)))
    
    markerDB <- markerDB[!is.na(markerDB$species) &
                             !is.na(markerDB$Symbol) &
                             !is.na(markerDB$PMID), ]
    
    markerDB$species <- tolower(markerDB$species)
    species <- tolower(species)
    
    if("all" %in% tolower(species)) {
        species <- speciesAll
    }
    
    
    if (!all(species %in% speciesAll)) {
        stop("Species", 
             paste(species[!species %in% speciesAll], collapse = ", "), 
             " not in the marker database.",
             " To explore the available species,",
             " use the showCellTaxoSpecies() function.")
    }
    
    
    markerDB2Use <- markerDB[markerDB$species %in% species, 
                             c("species", "cell_name", 
                               "Symbol", "PMID")]
    
    markerDB2Use$cellInfo <- paste(markerDB2Use$species, 
                                   markerDB2Use$cell_name, sep = ":")
    
    pmidCounts <- markerDB2Use %>% 
        group_by(cellInfo, Symbol) %>% 
        summarise(count = n_distinct(PMID))
    
    # Spread the data into a matrix format
    pmidCounts <- pmidCounts  %>% 
        tidyr::spread(Symbol, count)
    cellInfo <- pmidCounts$cellInfo
    
    pmidCounts <- as.matrix(pmidCounts[, -1])
    row.names(pmidCounts) <- cellInfo
    pmidCounts[is.na(pmidCounts)] <- 0
    
    markerEvi <- log(pmidCounts + pseudoCount, base = logBase)
    
    pmidCountsCellInfo <- markerDB2Use %>% 
        group_by(cellInfo) %>% 
        summarise(count = n_distinct(PMID))
    
    pmidCountsCellInfo <- 
        pmidCountsCellInfo[match(rownames(pmidCounts), 
                                 pmidCountsCellInfo$cellInfo), ]
    
    markerCountsCellInfo <- markerDB2Use %>% 
        group_by(cellInfo) %>% 
        summarise(count = n_distinct(Symbol))
    
    markerCountsCellInfo <- 
        markerCountsCellInfo[match(rownames(pmidCounts), 
                                   markerCountsCellInfo$cellInfo), ]
    
    cellInfoEvi <- log(pmidCountsCellInfo$count * markerCountsCellInfo$count +
                           pseudoCount, base = logBase)
    names(cellInfoEvi) <- rownames(pmidCounts)
    
    weight <- markerEvi * cellInfoEvi
    
    return(weight)
}


R/cellTypeAnnotator.R
#' Cell Type Annotation based on CellMarker 2.0 Database
#' 
#' This function assigns cell types to cell clusters of single-cell RNA-seq data based on differential gene expression information and marker gene evidence calculated using the CellMarker 2.0 database.
#'
#' @param clusterDEGs A data frame containing information about differentially expressed genes for each cell cluster. It should have the following columns:
#' \describe{
#'   \item{cluster}{The cluster identifier or label.}
#'   \item{gene}{The gene symbol.}
#'   \item{p_val}{The p-value representing the significance of differential expression for the gene in the cluster.}
#'   \item{avg_log2FC}{The average log fold change indicating the magnitude of differential expression for the gene in the cluster.}
#' } 
#' @param species A character vector specifying the species to consider. Currently it supports "human" and "mouse". Use "all" to include all species.
#' @param tissueClass A character vector specifying the broad tissue classes to consider (case insensitive). Default is "all" that includes all tissue classes. To explore the available tissue classes, use the \code{showCellMarkerTissue()} function.
#' @param tissueType A character vector specifying the specific tissue types to consider within the chosen tissue class (case insensitive). Default is "all" that includes all tissue types. "tissueType" represents the specific subtypes within "tissueClass", For instance, when considering blood as "tissueClass", examples of blood tissue types could be "Peripheral blood", "Serum", "Blood vessel", and so on. To explore the available tissue types, use the \code{showCellMarkerTissue()} function.
#' @param cellState A character vector specifying the cell states to consider. Valid options are "normal" and "cancer". Default is both.
#' @param markerSource A character vector specifying the marker gene sources to consider. Valid options are "experiment", "review", "single-cell sequencing", and "company". Default is all.
#' @param outFile The path where the xlsx file containing the annotation results will be saved.
#' @param clusters A character vector specifying the clusters to analyze. By default, all clusters are considered (set to "all").
#' @param cutoffLFC The log fold change cutoff value for the column "avg_log2FC" in the \code{clusterDEGs} data frame. Only cell type marker genes with an average log fold change greater than or equal to this value are considered as supporting evidence for the corresponding cell type. Default is 1, which corresponds to a two-fold change in average.
#' @param cutoffPval The p-value cutoff for the column "p_val" in the \code{clusterDEGs} data frame. Only cell type marker genes with a p-value less than or equal to this value are considered as supporting evidence for the corresponding cell type. Default is 0.05.
#' @param topCell The maximum number of top-ranked cell types to report. Default is 5.
#' @param topMarker The maximum number of top-ranked marker genes contributing to the cell type annotation to report for each cell type. Default is 5.
#' @param panelty The penalty weight assigned to cell type marker genes that are not a detected marker of the cluster. A "detected marker" means that the gene meets both the cutoffPval and paneltyCutoffLFC thresholds. A higher absolute value of the penalty means a more severe punishment for cell types with some of the marker genes that are not overexpressed in the cluster. The default value is -0.1.
#' @param paneltyCutoffLFC The log fold change cutoff value for including genes in the penalty calculation. Cell types with marker genes having an average log fold change less than this value are penalized. The default value is the same as "cutoffLFC". The value of "paneltyCutoffLFC" should be less than or equal to "cutoffLFC".
#'
#' @return A list containing the annotation results for each cluster. Each element in the list corresponds to a cluster and includes a data frame. The data frame has annotated cell types as row names and two columns: "score" (representing the final score for the cell type) and "topMarkers" (containing the top marker genes and their scores).
#' @export
#'
#' @examples
#' # Example usage
#' data(pbmc3kDEGs)
#' annoRes <- cellMarkerAnno(clusterDEGs = pbmc3kDEGs, 
#'                           species = "human", 
#'                           tissueClass = "Blood",
#'                           tissueType = "Peripheral blood",
#'                           cutoffLFC = 0.585)
#'
#'
#' @seealso
#' \code{\link{showCellMarkerTissue}}, \code{\link{cellTaxoAnno}}
#'
cellMarkerAnno <- function(clusterDEGs, 
                           species="human",
                           tissueClass="all",
                           tissueType="all",
                           cellState=c("normal", "cancer"),
                           markerSource=c("Experiment", 
                                          "Review",
                                          "Single-cell sequencing",
                                          "Company"),
                           outFile=NA,
                           clusters="all",
                           cutoffLFC=1,
                           cutoffPval=0.05,
                           topCell=5,
                           topMarker=5,
                           panelty=-0.1,
                           paneltyCutoffLFC=cutoffLFC) {
    
    stopifnot(!missing(clusterDEGs))
    markerDB <- cellMarkerDB
    
    markerEviMat <- cellMarkerSupport(markerDB = markerDB, 
                                      species = species, 
                                      tissueClass =  tissueClass,
                                      tissueType = tissueType,
                                      cellState = cellState,
                                      markerSource = markerSource,
                                      logBase = exp(1),
                                      pseudoCount = 1)
    
    cellTypeRes <- cellTypeIdentifier(clusterDEGs = clusterDEGs,
                                      markerEviMat = markerEviMat,
                                      clusters=clusters,
                                      cutoffLFC=cutoffLFC,
                                      cutoffPval=cutoffPval,
                                      topCell=topCell,
                                      topMarker=topMarker,
                                      panelty=panelty,
                                      paneltyCutoffLFC=paneltyCutoffLFC)
    
    if (!is.na(outFile)) {
        saveCellTypeRes(cellTypeRes=cellTypeRes, path=outFile)
    }
    
    return(cellTypeRes)
}
    



#' Cell Type Annotation based on Cell Taxonomy Database
#' 
#' This function assigns cell types to cell clusters of single-cell RNA-seq data based on differential gene expression information and marker gene evidence calculated using the Cell Taxonomy database.
#'
#' @param clusterDEGs A data frame containing information about differentially expressed genes for each cell cluster. It should have the following columns:
#' \describe{
#'   \item{cluster}{The cluster identifier or label.}
#'   \item{gene}{The gene symbol.}
#'   \item{p_val}{The p-value representing the significance of differential expression for the gene in the cluster.}
#'   \item{avg_log2FC}{The average log fold change indicating the magnitude of differential expression for the gene in the cluster.}
#' } 
#' @param species A character vector specifying the species to consider. Currently it contains data on 34 species. Use "all" to include all species. Default is "Homo sapiens". To explore the available species, use the \code{showCellTaxoSpecies()} function.
#' @param outFile The path where the xlsx file containing the annotation results will be saved.
#' @param clusters A character vector specifying the clusters to analyze. By default, all clusters are considered (set to "all").
#' @param cutoffLFC The log fold change cutoff value for the column "avg_log2FC" in the \code{clusterDEGs} data frame. Only cell type marker genes with an average log fold change greater than or equal to this value are considered as supporting evidence for the corresponding cell type. Default is 1, which corresponds to a two-fold change in average.
#' @param cutoffPval The p-value cutoff for the column "p_val" in the \code{clusterDEGs} data frame. Only cell type marker genes with a p-value less than or equal to this value are considered as supporting evidence for the corresponding cell type. Default is 0.05.
#' @param topCell The maximum number of top-ranked cell types to report. Default is 5.
#' @param topMarker The maximum number of top-ranked marker genes contributing to the cell type annotation to report for each cell type. Default is 5.
#' @param panelty The penalty weight assigned to cell type marker genes that are not a detected marker of the cluster. A "detected marker" means that the gene meets both the cutoffPval and paneltyCutoffLFC thresholds. A higher absolute value of the penalty means a more severe punishment for cell types with some of the marker genes that are not overexpressed in the cluster. The default value is -0.1.
#' @param paneltyCutoffLFC The log fold change cutoff value for including genes in the penalty calculation. Cell types with marker genes having an average log fold change less than this value are penalized. The default value is the same as "cutoffLFC". The value of "paneltyCutoffLFC" should be less than or equal to "cutoffLFC".
#'
#' @return A list containing the annotation results for each cluster. Each element in the list corresponds to a cluster and includes a data frame. The data frame has annotated cell types as row names and two columns: "score" (representing the final score for the cell type) and "topMarkers" (containing the top marker genes and their scores).
#' @export
#'
#' @examples
#' # Example usage
#' data(pbmc3kDEGs)
#' annoRes <- cellTaxoAnno(clusterDEGs = pbmc3kDEGs, 
#'                         species = "Homo sapiens",
#'                         cutoffLFC = 0.585,
#'                         clusters = "all")
#'
#'
#' @seealso
#' \code{\link{showCellTaxoSpecies}}, \code{\link{cellMarkerAnno}}
#'

cellTaxoAnno <- function(clusterDEGs,
                         species="Homo sapiens",
                         outFile=NA,
                         clusters="all",
                         cutoffLFC=1,
                         cutoffPval=0.05,
                         topCell=5,
                         topMarker=5,
                         panelty=-0.1,
                         paneltyCutoffLFC=cutoffLFC) {
    
    
    stopifnot(!missing(clusterDEGs))
    markerDB <- cellTaxoDB
    
    markerEviMat <- cellTaxonomySupport(markerDB = markerDB,
                                        species=species,
                                        logBase = exp(1),
                                        pseudoCount = 1)
    
    cellTypeRes <- cellTypeIdentifier(clusterDEGs = clusterDEGs,
                                      markerEviMat = markerEviMat,
                                      clusters=clusters,
                                      cutoffLFC=cutoffLFC,
                                      cutoffPval=cutoffPval,
                                      topCell=topCell,
                                      topMarker=topMarker,
                                      panelty=panelty,
                                      paneltyCutoffLFC=paneltyCutoffLFC)
    
    if (!is.na(outFile)) {
        saveCellTypeRes(cellTypeRes=cellTypeRes, path=outFile)
    }
    
    return(cellTypeRes)
}